SimulationAgent/Simulation/DeviceStatusLogic/UpdateReportedProperties.cs

```diff
@@ -0,0 +1,172 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Models;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Diagnostics;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.SimulationAgent.Exceptions;
+using System.Threading;
+
+namespace Microsoft.Azure.IoTSolutions.DeviceSimulation.SimulationAgent.Simulation.DeviceStatusLogic
+{
+
+    public class UpdateReportedProperties : IDeviceStatusLogic
+    {
+        // When connecting to IoT Hub, timeout after 10 seconds
+        private static readonly TimeSpan connectionTimeout = TimeSpan.FromSeconds(10);
+
+        private IDevices devices;
+        private string deviceId;
+        private DeviceModel deviceModel;
+
+        private readonly ILogger log;
+
+        // Ensure that setup is called once and only once (which helps also detecting thread safety issues)
+        private bool setupDone = false;
+
+        public UpdateReportedProperties(
+            IDevices devices,
+            ILogger logger)
+        {
+            this.log = logger;
+            this.devices = devices;
+        }
+
+        public void Run(object context)
+        {
+            try { 
+
+                this.ValidateSetup();
+
+                var actor = (IDeviceActor)context;
+                if (actor.CancellationToken.IsCancellationRequested)
+                {
+                    actor.Stop();
+                    return;
+                }
+
+                //TODO: Here we should pause the timer in case the device takes too long to pull from the hub
+
+                this.log.Debug(
+                    "Checking for desired property updates & updated reported properties",
+                    () => new { this.deviceId, deviceState = actor.DeviceState
+                    });
+
+                // Get device
+                var device = this.GetDevice(actor.CancellationToken);
+                lock (actor.DeviceState)
+                {
+                    // check for differences between reported/desired properties,
+                    // update reported properties with any state changes (either from desired prop changes, methods, etc.)
+                    if (this.ChangeTwinPropertiesToMatchDesired(device, actor.DeviceState)
+                        || this.ChangeTwinPropertiesToMatchActorState(device, actor.DeviceState))
```

Comment:
TODO (future): the device model should define whether the local state or the desired state wins, i.e. where is the master value

Reply:
https://github.com/Azure/device-simulation-dotnet/issues/76